internal/controller/pipelinerollout_progressive.go

```diff
@@ -0,0 +1,159 @@
+		}
+	}
+
+	done, err := r.processUpgradingPipelineStatus(ctx, pipelineRollout, existingPipelineDef)
+	if err != nil {
+		return false, err
+	}
+
+	return done, nil
+}
+
+func (r *PipelineRolloutReconciler) makeUpgradingPipelineDefinition(
+	ctx context.Context,
+	pipelineRollout *apiv1.PipelineRollout,
+) (*kubernetes.GenericObject, error) {
+	pipelineName, err := r.getPipelineName(ctx, pipelineRollout, string(common.LabelValueUpgradeInProgress))
+	if err != nil {
+		return nil, err
+	}
+
+	labels, err := pipelineLabels(pipelineRollout, string(common.LabelValueUpgradeInProgress))
+	if err != nil {
+		return nil, err
+	}
+
+	return r.makePipelineDefinition(pipelineRollout, pipelineName, labels)
+}
+
+func (r *PipelineRolloutReconciler) processUpgradingPipelineStatus(
+	ctx context.Context,
+	pipelineRollout *apiv1.PipelineRollout,
+	existingPipelineDef *kubernetes.GenericObject,
+) (bool, error) {
+	numaLogger := logger.FromContext(ctx)
+
+	pipelineDef, err := r.makeUpgradingPipelineDefinition(ctx, pipelineRollout)
+	if err != nil {
+		return false, err
+	}
+
+	// Get existing upgrading Pipeline
+	existingUpgradingPipelineDef, err := kubernetes.GetCR(ctx, r.restConfig, pipelineDef, "pipelines")
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			numaLogger.WithValues("pipelineDefinition", *pipelineDef).
+				Warn("Pipeline not found. Unable to process status during this reconciliation.")
+		} else {
+			return false, fmt.Errorf("error getting Pipeline for status processing: %v", err)
+		}
+	}
+
+	pipelineStatus, err := kubernetes.ParseStatus(existingUpgradingPipelineDef)
+	if err != nil {
+		return false, fmt.Errorf("failed to parse Pipeline Status from pipeline CR: %+v, %v", existingUpgradingPipelineDef, err)
+	}
+
+	pipelinePhase := numaflowv1.PipelinePhase(pipelineStatus.Phase)
+	if pipelinePhase == numaflowv1.PipelinePhaseFailed {
+		err = r.updatePipelineLabel(ctx, r.restConfig, existingUpgradingPipelineDef, "")
+		if err != nil {
+			return false, err
+		}
+		pipelineRollout.Status.MarkPipelineProgressiveUpgradeFailed("New Pipeline Failed", pipelineRollout.Generation)
+		return false, nil
+	} else if pipelinePhase == numaflowv1.PipelinePhaseRunning {
+		// Label the new pipeline as promoted and then remove the label from the old pipeline,
+		// since per PipelineRollout is reconciled only once at a time, we do not
+		// need to worry about consistency issue.
+		err = r.updatePipelineLabel(ctx, r.restConfig, existingUpgradingPipelineDef, string(common.LabelValueUpgradePromoted))
+		if err != nil {
+			return false, err
+		}
+
+		err = r.updatePipelineLabel(ctx, r.restConfig, existingPipelineDef, "")
+		if err != nil {
+			return false, err
+		}
+
+		pipelineRollout.Status.MarkPipelineProgressiveUpgradeSucceeded("New Pipeline Running", pipelineRollout.Generation)
+		pipelineRollout.Status.MarkDeployed(pipelineRollout.Generation)
+
+		// TODO: pause old pipeline
+		return true, nil
+	} else {
```

Comment:
I like this logic - it handles the race condition of the existing upgrading Pipeline entering "Failed" state or "Running" state at the same time as there's a new update in the PipelineRollout spec. In that case we mark as "promoted" or failed, and then in a subsequent reconciliation actually process the latest desired PipelineRollout spec.

Reply:
Although, just thinking that theoretically, in the case that there are 2 back to back updates prior to the pipeline entering "Running", we may notice it "Running" and mark it as such prior to Numaflow having reconciled the very latest update that we wrote. That's unlikely in this current state to happen, but once the Progressive process becomes longer, we may need to be cognizant of the Pipeline's ObservedGeneration.

Reply:
In the future, we may even want to stop the current upgrade which is in progress and have a whole new one if this happens.

Reply:
I don't think stopping the current upgrade is needed, as we always check if there is a new spec change needs to happen by compare the rollout spec and the `promoted` pipeline spec (which is a new pipeline).

Reply:
Well, in the future when we have "Progressive", I'm thinking say we're analyzing how successfully that new Pipeline is running - I don't want to be in a situation in which we replace the new Pipeline manifest and we still use the original data. That's why I thought one possible way to deal with it is to stop the current upgrade and start up a new one.

Reply:
ah got it, that could be needed in that case